### PR TITLE
646 field contains string

### PR DESCRIPTION
**Pull request by Mefpef**
Type pull request title with short description (#646 )

646 field contains string

add contains string plugin and register it
----

**Please** following this checklist to help us incorporate your contribution quickly and easily:

 - [x] Give the same name to the pull request as the name of ticket 
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Test your code throughly before PR.
 
If your pull request is about ~20 lines of code add license information in your code in form of comment. i.e. # License: MIT, Author:  Your name.

To make clear that you license your contribution under the MIT you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [MIT license](https://opensource.org/licenses/MIT)

PRs without ticked MIT license will bot be accepted - Sorry.

### CODE REVIEW

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+from tracardi.service.plugin.domain.config import PluginConfig
+
+from tracardi.service.plugin.domain.register import Plugin, Spec, MetaData, Documentation, PortDoc, Form, FormGroup, \
+    FormField, FormComponent
+from tracardi.service.plugin.runner import ActionRunner
+from tracardi.service.plugin.domain.result import Result
+
+
+class Config(PluginConfig):
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:11]: I'd move it to another directory, should be somewhere in operations or strings section of /action/v1, with config separated in .mode.config file

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    field: str
+    prefix: str
+
+    @validator("prefix")
+    def if_prefix_is_empty(cls, value):
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:16]: def validate_prefix(cls, value):

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+
+class ContainsStringAction(ActionRunner):
+
+    def __init__(self, **kwargs):
+        self.config = validate(kwargs)
+
+    async def run(self, payload: dict, in_edge=None):
+        dot = self._get_dot_accessor(payload)
+        if self.config.prefix in dot[self.config.field]:
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:33]: error prone, you should first check if dot[self.config.field] is iterable, and log, then throw an exception with an explanation

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        spec=Spec(
+            module=__name__,
+            className='ContainsStringAction',
+            inputs=["payload"],
+            outputs=["true", "false"],
+            version='0.7.2',
+            license="MIT",
+            author="Mateusz Zitaruk",
+            init={
+                "field": "",
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:51]: should default to None

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+from tracardi.service.plugin.domain.config import PluginConfig
+
+from tracardi.service.plugin.domain.register import Plugin, Spec, MetaData, Documentation, PortDoc, Form, FormGroup, \
+    FormField, FormComponent
+from tracardi.service.plugin.runner import ActionRunner
+from tracardi.service.plugin.domain.result import Result
+
+
+class Config(PluginConfig):
+    field: str
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:12]: should be validated

Comment (DawidekZagajnik) [reply tracardi/process_engine/action/v1/contains_string_action.py:12]: if not empty as well

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            },
+            manual="contains_string_action",
+            form=Form(
+                groups=[
+                    FormGroup(
+                        name="Field contains string plugin configuration",
+                        fields=[
+                            FormField(
+                                id="field",
+                                name="Type string or reference to string which you want to check.",
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:62]: type string or path to the string that you want to check.

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                                component=FormComponent(
+                                    type="dotPath",
+                                    props={
+                                        "label": "Payload field",
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:66]: label: path

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+                                        "defaultSourceValue": "event"
+                                    }
+                                )
+                            ),
+                            FormField(
+                                id="prefix",
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:72]: rename it to substring

Comment (DawidekZagajnik) [reply tracardi/process_engine/action/v1/contains_string_action.py:72]: basically everywhere

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+                                )
+                            )
+                        ]
+                    )
+                ]
+            )
+
+        ),
+        metadata=MetaData(
+            name='Contain string',
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:89]: Contains

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            desc='Checks if field contains defined string.',
+            icon='question',
+            group=["Flow control"],
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:92]: more like strings or operations

File: tracardi/process_engine/action/v1/contains_string_action.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+            documentation=Documentation(
+                inputs={
+                    "payload": PortDoc(desc="This port takes payload object.")
+                },
+                outputs={"true": PortDoc(desc="This port returns payload if field contains defined string."),
```
Comment (DawidekZagajnik) [inline tracardi/process_engine/action/v1/contains_string_action.py:97]: formatting

**Review by DawidekZagajnik — CHANGES_REQUESTED** [review]
refactoring and formatting needed

**Review by atompie — APPROVED** [review]

**Pull request merged**